File: sdk/ctv/android-tv/src/main/kotlin/com/rivalapex/androidtv/mediation/NoFillTracker.kt
package com.rivalapex.androidtv.mediation

import java.util.concurrent.ConcurrentHashMap
import java.util.concurrent.atomic.AtomicInteger
import java.util.concurrent.atomic.AtomicLong
import java.util.Calendar
import java.util.concurrent.CopyOnWriteArrayList

/**
 * Reasons for ad no-fill on Android TV
 */
enum class NoFillReason(val value: String) {
    TIMEOUT("timeout"),
    NO_INVENTORY("no_inventory"),
    NETWORK_ERROR("network_error"),
    POLICY_VIOLATION("policy_violation"),
    FREQUENCY_CAP("frequency_cap"),
    GEOGRAPHIC_RESTRICTION("geographic_restriction"),
    BUDGET_EXHAUSTED("budget_exhausted"),
    MALFORMED_RESPONSE("malformed_response"),
    SERVER_ERROR("server_error"),
    CTV_NOT_SUPPORTED("ctv_not_supported"),
    FORMAT_MISMATCH("format_mismatch"),
    UNKNOWN("unknown")
}

/**
 * A single no-fill event for Android TV
 */
data class NoFillEvent(
    val sourceId: String,
    val placementId: String,
    val reason: NoFillReason,
    val timestamp: Long = System.currentTimeMillis(),
    val latencyMs: Long = 0,
    val adFormat: String? = null,
    val metadata: Map<String, String> = emptyMap()
)

/**
 * Statistics for no-fill events
 */
data class NoFillStats(
    val totalNoFills: Int,
    val noFillRate: Double,
    val averageLatencyMs: Double,
    val topReasons: Map<NoFillReason, Int>,
    val topSources: Map<String, Int>,
    val topPlacements: Map<String, Int>,
    val formatBreakdown: Map<String, Int>,
    val hourlyBreakdown: Map<Int, Int>
)

/**
 * Detected no-fill pattern
 */
data class NoFillPattern(
    val type: PatternType,
    val severity: Severity,
    val description: String,
    val detectedAt: Long = System.currentTimeMillis(),
    val affectedSourceId: String? = null,
    val affectedPlacementId: String? = null
) {
    enum class PatternType {
        ELEVATED_RATE,
        SOURCE_SPECIFIC,
        PLACEMENT_SPECIFIC,
        FORMAT_SPECIFIC,
        CONSECUTIVE_FAILURES
    }
    
    enum class Severity {
        LOW, MEDIUM, HIGH, CRITICAL
    }
}

/**
 * Configuration for NoFillTracker
 */
data class NoFillTrackerConfig(
    val maxEventsRetained: Int = 5_000,
    val maxRetentionHours: Int = 12,
    val elevatedRateThreshold: Double = 0.5,
    val patternDetectionEnabled: Boolean = true,
    val consecutiveFailureThreshold: Int = 3
)

/**
 * Pattern listener interface
 */
fun interface NoFillPatternListener {
    fun onPatternDetected(pattern: NoFillPattern)
}

/**
 * Tracks no-fill events for Android TV analytics
 */
object NoFillTracker {
    
    private var config = NoFillTrackerConfig()
    private val events = CopyOnWriteArrayList<NoFillEvent>()
    private val lock = Object()
    
    private val totalNoFills = AtomicInteger(0)
    private val totalLatencyMs = AtomicLong(0)
    
    private val hourlyNoFills = ConcurrentHashMap<Int, AtomicInteger>()
    private val noFillsBySource = ConcurrentHashMap<String, AtomicInteger>()
    private val noFillsByPlacement = ConcurrentHashMap<String, AtomicInteger>()
    private val noFillsByReason = ConcurrentHashMap<NoFillReason, AtomicInteger>()
    private val noFillsByFormat = ConcurrentHashMap<String, AtomicInteger>()
    
    private val consecutiveNoFillsBySource = ConcurrentHashMap<String, AtomicInteger>()
    private val detectedPatterns = CopyOnWriteArrayList<NoFillPattern>()
    private val patternListeners = CopyOnWriteArrayList<NoFillPatternListener>()
    
    /**
     * Record a no-fill event
     */
    fun recordNoFill(
        sourceId: String,
        placementId: String,
        reason: NoFillReason,
        latencyMs: Long = 0,
        adFormat: String? = null,
        metadata: Map<String, String> = emptyMap()
    ) {
        val event = NoFillEvent(
            sourceId = sourceId,
            placementId = placementId,
            reason = reason,
            latencyMs = latencyMs,
            adFormat = adFormat,
            metadata = metadata
        )
        
        synchronized(lock) {
            events.add(event)
            totalNoFills.incrementAndGet()
            totalLatencyMs.addAndGet(latencyMs)
            
            noFillsBySource.getOrPut(sourceId) { AtomicInteger(0) }.incrementAndGet()
            noFillsByPlacement.getOrPut(placementId) { AtomicInteger(0) }.incrementAndGet()
            noFillsByReason.getOrPut(reason) { AtomicInteger(0) }.incrementAndGet()
            
            adFormat?.let { format ->
                noFillsByFormat.getOrPut(format) { AtomicInteger(0) }.incrementAndGet()
            }
            
            val calendar = Calendar.getInstance()
            calendar.timeInMillis = event.timestamp
            val hour = calendar.get(Calendar.HOUR_OF_DAY)
            hourlyNoFills.getOrPut(hour) { AtomicInteger(0) }.incrementAndGet()
            
            consecutiveNoFillsBySource.getOrPut(sourceId) { AtomicInteger(0) }.incrementAndGet()
            
            cleanupIfNeeded()
            
            if (config.patternDetectionEnabled) {
                detectPatternsForEvent(event)
            }
        }
    }
    
    /**
     * Record a successful fill
     */
    fun recordFill(sourceId: String) {
        consecutiveNoFillsBySource[sourceId]?.set(0)
    }
    
    /**
     * Get current statistics
     */
    fun getStats(): NoFillStats {
        val total = totalNoFills.get()
        val avgLatency = if (total > 0) {
            totalLatencyMs.get().toDouble() / total
        } else 0.0
        
        val oneHourAgo = System.currentTimeMillis() - 3_600_000
        val recentCount = events.count { it.timestamp >= oneHourAgo }
        val ratePerMinute = recentCount.toDouble() / 60.0
        
        return NoFillStats(
            totalNoFills = total,
            noFillRate = ratePerMinute,
            averageLatencyMs = avgLatency,
            topReasons = noFillsByReason.mapValues { it.value.get() },
            topSources = noFillsBySource.mapValues { it.value.get() },
            topPlacements = noFillsByPlacement.mapValues { it.value.get() },
            formatBreakdown = noFillsByFormat.mapValues { it.value.get() },
            hourlyBreakdown = hourlyNoFills.mapValues { it.value.get() }
        )
    }
    
    /**
     * Get no-fills by ad format
     */
    fun getNoFillsByFormat(): Map<String, Int> {
        return noFillsByFormat.mapValues { it.value.get() }
    }
    
    /**
     * Get detected patterns
     */
    fun getDetectedPatterns(): List<NoFillPattern> {
        return detectedPatterns.toList()
    }
    
    /**
     * Add pattern listener
     */
    fun addPatternListener(listener: NoFillPatternListener) {
        patternListeners.add(listener)
    }
    
    /**
     * Remove pattern listener
     */
    fun removePatternListener(listener: NoFillPatternListener) {
        patternListeners.remove(listener)
    }
    
    /**
     * Get recent events
     */
    fun getRecentEvents(count: Int = 50): List<NoFillEvent> {
        return events.takeLast(count)
    }
    
    /**
     * Clear all data
     */
    fun clear() {
        synchronized(lock) {
            events.clear()
            totalNoFills.set(0)
            totalLatencyMs.set(0)
            hourlyNoFills.clear()
            noFillsBySource.clear()
            noFillsByPlacement.clear()
            noFillsByReason.clear()
            noFillsByFormat.clear()
            consecutiveNoFillsBySource.clear()
            detectedPatterns.clear()
        }
    }
    
    /**
     * Update configuration
     */
    fun updateConfiguration(newConfig: NoFillTrackerConfig) {
        synchronized(lock) {
            config = newConfig
        }
    }
    
    private fun cleanupIfNeeded() {
        val cutoff = System.currentTimeMillis() - (config.maxRetentionHours * 3_600_000L)
        events.removeAll { it.timestamp < cutoff }
        
        while (events.size > config.maxEventsRetained) {
            events.removeAt(0)
        }
    }
    
    private fun detectPatternsForEvent(event: NoFillEvent) {
        // Check for consecutive failures
        val consecutive = consecutiveNoFillsBySource[event.sourceId]?.get() ?: 0
        if (consecutive >= config.consecutiveFailureThreshold) {
            val severity = when {
                consecutive >= 10 -> NoFillPattern.Severity.CRITICAL
                consecutive >= 5 -> NoFillPattern.Severity.HIGH
                else -> NoFillPattern.Severity.MEDIUM
            }
            val pattern = NoFillPattern(
                type = NoFillPattern.PatternType.CONSECUTIVE_FAILURES,
                severity = severity,
                description = "Source ${event.sourceId} has $consecutive consecutive no-fills",
                affectedSourceId = event.sourceId
            )
            addPattern(pattern)
        }
        
        // Check for format-specific issues
        event.adFormat?.let { format ->
            val formatTotal = noFillsByFormat[format]?.get() ?: 0
            if (formatTotal > 10) {
                val total = maxOf(1, totalNoFills.get())
                val formatRate = formatTotal.toDouble() / total
                if (formatRate > 0.5) {
                    val pattern = NoFillPattern(
                        type = NoFillPattern.PatternType.FORMAT_SPECIFIC,
                        severity = NoFillPattern.Severity.HIGH,
                        description = "Format '$format' accounts for ${(formatRate * 100).toInt()}% of no-fills"
                    )
                    addPattern(pattern)
                }
            }
        }
    }
    
    private fun addPattern(pattern: NoFillPattern) {
        val recentCutoff = System.currentTimeMillis() - 300_000
        val isDuplicate = detectedPatterns.any { existing ->
            existing.type == pattern.type &&
            existing.affectedSourceId == pattern.affectedSourceId &&
            existing.detectedAt > recentCutoff
        }
        
        if (!isDuplicate) {
            detectedPatterns.add(pattern)
            
            val oneDayAgo = System.currentTimeMillis() - 86_400_000
            detectedPatterns.removeAll { it.detectedAt < oneDayAgo }
            
            patternListeners.forEach { it.onPatternDetected(pattern) }
        }
    }
}

File: sdk/ctv/tvos/Sources/CTVSDK/Mediation/NoFillTracker.swift
import Foundation

// MARK: - NoFillReason

/// Reasons for ad no-fill on CTV
public enum NoFillReason: String, Codable {
    case timeout = "timeout"
    case noInventory = "no_inventory"
    case networkError = "network_error"
    case policyViolation = "policy_violation"
    case frequencyCap = "frequency_cap"
    case geographicRestriction = "geographic_restriction"
    case budgetExhausted = "budget_exhausted"
    case malformedResponse = "malformed_response"
    case serverError = "server_error"
    case ctvNotSupported = "ctv_not_supported"
    case formatMismatch = "format_mismatch"
    case unknown = "unknown"
}

// MARK: - NoFillEvent

/// A single no-fill event for CTV
public struct NoFillEvent: Codable {
    public let sourceId: String
    public let placementId: String
    public let reason: NoFillReason
    public let timestamp: Date
    public let latencyMs: Int64
    public let adFormat: String?
    public let metadata: [String: String]
    
    public init(
        sourceId: String,
        placementId: String,
        reason: NoFillReason,
        timestamp: Date = Date(),
        latencyMs: Int64 = 0,
        adFormat: String? = nil,
        metadata: [String: String] = [:]
    ) {
        self.sourceId = sourceId
        self.placementId = placementId
        self.reason = reason
        self.timestamp = timestamp
        self.latencyMs = latencyMs
        self.adFormat = adFormat
        self.metadata = metadata
    }
}

// MARK: - NoFillStats

/// Statistics for no-fill events
public struct NoFillStats {
    public let totalNoFills: Int
    public let noFillRate: Double
    public let averageLatencyMs: Double
    public let topReasons: [NoFillReason: Int]
    public let topSources: [String: Int]
    public let topPlacements: [String: Int]
    public let formatBreakdown: [String: Int]
    public let hourlyBreakdown: [Int: Int]
}

// MARK: - NoFillPattern

/// Detected no-fill pattern
public struct NoFillPattern {
    public let type: PatternType
    public let severity: Severity
    public let description: String
    public let detectedAt: Date
    public let affectedSourceId: String?
    public let affectedPlacementId: String?
    
    public enum PatternType: String {
        case elevatedRate = "elevated_rate"
        case sourceSpecific = "source_specific"
        case placementSpecific = "placement_specific"
        case formatSpecific = "format_specific"
        case consecutiveFailures = "consecutive_failures"
    }
    
    public enum Severity: String {
        case low = "low"
        case medium = "medium"
        case high = "high"
        case critical = "critical"
    }
}

// MARK: - NoFillTracker

/// Tracks no-fill events for CTV analytics
public final class NoFillTracker {
    
    // MARK: - Singleton
    
    public static let shared = NoFillTracker()
    
    // MARK: - Configuration
    
    public struct Configuration {
        public var maxEventsRetained: Int = 5_000 // Less memory for tvOS
        public var maxRetentionHours: Int = 12 // Shorter retention for CTV
        public var elevatedRateThreshold: Double = 0.5
        public var patternDetectionEnabled: Bool = true
        public var consecutiveFailureThreshold: Int = 3 // More sensitive for CTV
        
        public static let `default` = Configuration()
    }
    
    // MARK: - Properties
    
    private var config: Configuration
    private var events: [NoFillEvent] = []
    private let lock = NSLock()
    
    private var totalNoFills: Int = 0
    private var totalLatencyMs: Int64 = 0
    
    private var hourlyNoFills: [Int: Int] = [:]
    private var noFillsBySource: [String: Int] = [:]
    private var noFillsByPlacement: [String: Int] = [:]
    private var noFillsByReason: [NoFillReason: Int] = [:]
    private var noFillsByFormat: [String: Int] = [:]
    
    private var consecutiveNoFillsBySource: [String: Int] = [:]
    private var detectedPatterns: [NoFillPattern] = []
    private var patternListeners: [(NoFillPattern) -> Void] = []
    
    // MARK: - Initialization
    
    private init(config: Configuration = .default) {
        self.config = config
    }
    
    // MARK: - Public API
    
    /// Record a no-fill event
    public func recordNoFill(
        sourceId: String,
        placementId: String,
        reason: NoFillReason,
        latencyMs: Int64 = 0,
        adFormat: String? = nil,
        metadata: [String: String] = [:]
    ) {
        let event = NoFillEvent(
            sourceId: sourceId,
            placementId: placementId,
            reason: reason,
            latencyMs: latencyMs,
            adFormat: adFormat,
            metadata: metadata
        )
        
        lock.lock()
        defer { lock.unlock() }
        
        events.append(event)
        totalNoFills += 1
        totalLatencyMs += latencyMs
        
        noFillsBySource[sourceId, default: 0] += 1
        noFillsByPlacement[placementId, default: 0] += 1
        noFillsByReason[reason, default: 0] += 1
        
        if let format = adFormat {
            noFillsByFormat[format, default: 0] += 1
        }
        
        let hour = Calendar.current.component(.hour, from: event.timestamp)
        hourlyNoFills[hour, default: 0] += 1
        
        consecutiveNoFillsBySource[sourceId, default: 0] += 1
        
        cleanupIfNeeded()
        
        if config.patternDetectionEnabled {
            detectPatternsForEvent(event)
        }
    }
    
    /// Record a successful fill
    public func recordFill(sourceId: String) {
        lock.lock()
        defer { lock.unlock() }
        consecutiveNoFillsBySource[sourceId] = 0
    }
    
    /// Get current statistics
    public func getStats() -> NoFillStats {
        lock.lock()
        defer { lock.unlock() }
        
        let avgLatency = totalNoFills > 0 
            ? Double(totalLatencyMs) / Double(totalNoFills) 
            : 0.0
        
        let oneHourAgo = Date().addingTimeInterval(-3600)
        let recentEvents = events.filter { $0.timestamp >= oneHourAgo }
        let ratePerMinute = Double(recentEvents.count) / 60.0
        
        return NoFillStats(
            totalNoFills: totalNoFills,
            noFillRate: ratePerMinute,
            averageLatencyMs: avgLatency,
            topReasons: noFillsByReason,
            topSources: noFillsBySource,
            topPlacements: noFillsByPlacement,
            formatBreakdown: noFillsByFormat,
            hourlyBreakdown: hourlyNoFills
        )
    }
    
    /// Get no-fills by ad format
    public func getNoFillsByFormat() -> [String: Int] {
        lock.lock()
        defer { lock.unlock() }
        return noFillsByFormat
    }
    
    /// Get detected patterns
    public func getDetectedPatterns() -> [NoFillPattern] {
        lock.lock()
        defer { lock.unlock() }
        return detectedPatterns
    }
    
    /// Add pattern listener
    public func addPatternListener(_ listener: @escaping (NoFillPattern) -> Void) {
        lock.lock()
        defer { lock.unlock() }
        patternListeners.append(listener)
    }
    
    /// Get recent events
    public func getRecentEvents(count: Int = 50) -> [NoFillEvent] {
        lock.lock()
        defer { lock.unlock() }
        return Array(events.suffix(count))
    }
    
    /// Clear all data
    public func clear() {
        lock.lock()
        defer { lock.unlock() }
        
        events.removeAll()
        totalNoFills = 0
        totalLatencyMs = 0
        hourlyNoFills.removeAll()
        noFillsBySource.removeAll()
        noFillsByPlacement.removeAll()
        noFillsByReason.removeAll()
        noFillsByFormat.removeAll()
        consecutiveNoFillsBySource.removeAll()
        detectedPatterns.removeAll()
    }
    
    /// Update configuration
    public func updateConfiguration(_ config: Configuration) {
        lock.lock()
        defer { lock.unlock() }
        self.config = config
    }
    
    // MARK: - Private Methods
    
    private func cleanupIfNeeded() {
        let cutoff = Date().addingTimeInterval(-Double(config.maxRetentionHours) * 3600)
        events.removeAll { $0.timestamp < cutoff }
        
        if events.count > config.maxEventsRetained {
            let excess = events.count - config.maxEventsRetained
            events.removeFirst(excess)
        }
    }
    
    private func detectPatternsForEvent(_ event: NoFillEvent) {
        // Check for consecutive failures
        if let consecutive = consecutiveNoFillsBySource[event.sourceId],
           consecutive >= config.consecutiveFailureThreshold {
            let pattern = NoFillPattern(
                type: .consecutiveFailures,
                severity: consecutive >= 10 ? .critical : (consecutive >= 5 ? .high : .medium),
                description: "Source \(event.sourceId) has \(consecutive) consecutive no-fills",
                detectedAt: Date(),
                affectedSourceId: event.sourceId,
                affectedPlacementId: nil
            )
            addPattern(pattern)
        }
        
        // Check for format-specific issues
        if let format = event.adFormat {
            let formatTotal = noFillsByFormat[format] ?? 0
            if formatTotal > 10 {
                let formatRate = Double(formatTotal) / Double(max(1, totalNoFills))
                if formatRate > 0.5 {
                    let pattern = NoFillPattern(
                        type: .formatSpecific,
                        severity: .high,
                        description: "Format '\(format)' accounts for \(Int(formatRate * 100))% of no-fills",
                        detectedAt: Date(),
                        affectedSourceId: nil,
                        affectedPlacementId: nil
                    )
                    addPattern(pattern)
                }
            }
        }
    }
    
    private func addPattern(_ pattern: NoFillPattern) {
        let recentCutoff = Date().addingTimeInterval(-300)
        let isDuplicate = detectedPatterns.contains { existing in
            existing.type == pattern.type &&
            existing.affectedSourceId == pattern.affectedSourceId &&
            existing.detectedAt > recentCutoff
        }
        
        if !isDuplicate {
            detectedPatterns.append(pattern)
            
            let oneDayAgo = Date().addingTimeInterval(-86400)
            detectedPatterns.removeAll { $0.detectedAt < oneDayAgo }
            
            for listener in patternListeners {
                listener(pattern)
            }
        }
    }
}
